src/Native/Runtime/arm/PInvoke.S

```diff
@@ -29,47 +17,48 @@
 // The codegenerator must treat the callsite of this helper as GC triggering and generate the GC info for it.
 // Also, the codegenerator must ensure that there are no live GC references in callee saved registers.
 //
-NESTED_ENTRY RhpPInvoke, _TEXT, NoHandler
-	        PROLOG_PUSH "{r4-r7,lr}"
 
-	        str         lr, [r0, #OFFSETOF__PInvokeTransitionFrame__m_RIP]
-	        str         r7, [r0, #OFFSETOF__PInvokeTransitionFrame__m_FramePointer]
-	        str         sp, [r0, #OFFSETOF__PInvokeTransitionFrame__m_PreservedRegs]
+NESTED_ENTRY RhpPInvoke, _TEXT, NoHandler
+        str     lr, [r0, #OFFSETOF__PInvokeTransitionFrame__m_RIP]
+        str     r7, [r0, #OFFSETOF__PInvokeTransitionFrame__m_FramePointer]
+        str     sp, [r0, #OFFSETOF__PInvokeTransitionFrame__m_PreservedRegs]
 
-	        mov	        r5, r0
-	        // R0 = GetThread(), R6 - Trash Register
-	        INLINE_GETTHREAD r6
-	        str         r0, [r5, #OFFSETOF__PInvokeTransitionFrame__m_pThread]
-	        str         r5, [r0, #OFFSETOF__Thread__m_pTransitionFrame]
+        PROLOG_PUSH "{r5,lr}"
 
-	        ldr         r6, =C_FUNC(RhpTrapThreads)
-	        ldr         r6,	[r6]
-	        cbnz        r6, LOCAL_LABEL(InvokeRareTrapThread)
+        mov   r5, r0
+        // get TLS global variable address
+        INLINE_GET_TLS_GD_VAR tls_CurrentThread
+        str     r0, [r5, #OFFSETOF__PInvokeTransitionFrame__m_pThread]
+        str     r5, [r0, #OFFSETOF__Thread__m_pTransitionFrame]
+        ldr     r3, =C_FUNC(RhpTrapThreads)
+        ldr     r3, [r3]
 
-	        EPILOG_POP  "{r4-r7,pc}"
+        EPILOG_POP "{r5,lr}"
+        cbnz    r3, LOCAL_LABEL(InvokeRareTrapThread)
+        mov     pc, lr
```

Comment:
The ARM silicon has the specific epilog sequences optimized. I suspect that `mov pc, lr` is not among the optimized ones. I would be better to use the common epilog sequence here.

```
cbnz
EPILOG_POP "{r5,pc}
....
EPILOG_POP "{r5,lr}
b C_FUNC(RhpWaitForSuspend2)
```